pkg/artifactory/datasource_artifactory_file_test.go

```diff
@@ -1,58 +1,173 @@
+		}
+		return err
+	}
+
 	resource.Test(t, resource.TestCase{
-		PreCheck: func() {
-			testAccPreCheck(t)
-			client := getTestResty(t)
-			err := uploadTestFile(client, "../../samples/crash.zip", "example-repo-local/crash.zip", "application/zip")
-			if err != nil {
-				panic(err)
-			}
+		PreCheck:          downloadPreCheck(t, downloadPath, &localFileModTime),
+		ProviderFactories: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(script, downloadPath),
+				Check:  downloadCheck,
+			},
 		},
+	})
+}
+
+/*
+Negative test case on file download skip
+When file is present at output_path, checksum of files at output_path & repository path matches
+artifactory_file datasource will skip the download.
+*/
+func TestFileDownloadSkipCheck(t *testing.T) {
+	downloadPath := fmt.Sprintf("%s/crash.zip", t.TempDir())
+	localFileModTime := time.Time{}
+
+	// every instance of RT has this repo and file out-of-the-box
+	const script = `
+		data "artifactory_file" "example" {
+		  repository      = "example-repo-local"
+		  path            = "crash.zip"
+		  output_path     = "%s"
+		  force_overwrite = false
+		}
+	`
+
+	var skipDownloadCheck = func(state *terraform.State) error {
+		download := state.Modules[0].Resources["data.artifactory_file.example"].Primary.Attributes["output_path"]
+		downloadedFileStat, err := os.Stat(download)
+		if err != nil {
+			return err
+		}
+		downloadedFileModTime := downloadedFileStat.ModTime()
+		if downloadedFileModTime.After(localFileModTime) { //determines fresh download occurred during the test step
+			return fmt.Errorf("fresh download observed. Existing file modification time: %v. Fresh downloaded file modification time: %v", localFileModTime, downloadedFileModTime)
+		}
+		return nil
+	}
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:          downloadPreCheck(t, downloadPath, &localFileModTime),
 		ProviderFactories: testAccProviders,
 		Steps: []resource.TestStep{
 			{
-				Config: script,
-				Check: func(state *terraform.State) error {
-					download := state.Modules[0].Resources["data.artifactory_file.example"].Primary.Attributes["output_path"]
-					_, err := os.Stat(download)
-					if err != nil {
-						return err
-					}
-					verified, err := VerifySha256Checksum(download, "7a2489dd209d0acb72f7f11d171b418e65648b9cc96c6c351e00e22551fdd8f1")
-
-					if !verified {
-						return fmt.Errorf("%s checksum does not have expected checksum", download)
-					}
-					return err
-				},
+				Config: fmt.Sprintf(script, downloadPath),
+				Check:  skipDownloadCheck,
 			},
 		},
 	})
 }
+
+//Creates new directory tree if not exist
+func createNewDir(srcPath string) error {
+	_, err := os.Stat(srcPath)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			errMkDirAll := os.MkdirAll(srcPath, os.ModePerm)
+			if errMkDirAll != nil {
+				return errMkDirAll
+			}
+		} else {
+			return err
+		}
+	}
+	return nil
```

Comment:
This can be simplified a bit to reduce cyclomatic complexity.

```go
_, err := os.Stat(srcPath)
if err == nil {
  return nil
}
// if errors.Is can handle a nil error, then even the line above can go
if !errors.Is(err,os.ErrorNotExist) {
   return err
}
errMkDirAll := os.MkdirAll(srcPath, os.ModePerm)
if errMkDirAll != nil {
   return errMkDirAll
}
return nil
```

Notice how there is no nesting of conditions to figure out?
Generally speaking, nested branches in logic aren't good.

Reply:
improved